### ISSUE TITLE
[API Proposal]: FakeTimeProvider, option to allow going backwards

### ISSUE THREAD

**Issue by jjxtra** [dotnet/extensions#5072]
### Background and motivation

I'm using reflection to get around FakeTimeProvider inability to go backwards. Would be nice if a bool could be added to allow setting time earlier.

Thanks for considering.

### API Proposal

```csharp

public sealed class FakeTimeProvider : TimeProvider
{
    public bool AllowBackwardsTimeProgression { get; set; }
}
```

### API Usage

```csharp
// Fancy the value
FakeTimeProvider t = ...
t.AllowBackwardsTimeProgression = true;
t.SetUtcNow(new DateTimeOffset(1990, 1, 1, 1, 1, 1));
t.SetUtcNow(new DateTimeOffset(1984, 1, 1, 1, 1, 1));
```

### Risks

None I can think of. This new property will default to false.

**Comment by geeknoid** [comment]
@stephentoub I presume it's possible for the normal time provider's idea of current time to move backwards? And this presumably doesn't affect any outstanding timers (since they are created with relative time), and presumably wouldn't affect the value returned by GetTimestamp?

**Comment by jjxtra** [comment]
The internal property storing the time for FakeTimeProvider is called _now and is simply a DateTimeOffset value.

**Comment by evgenyfedorov2** [comment]
@geeknoid -can this issue be closed as fixed by https://github.com/dotnet/extensions/pull/5192 ?

**Comment by geeknoid** [comment]
@evgenyfedorov2 We need an API review to approve the new API, so we can remove the experimental attribute, and then we can close this issue.

**Comment by amadeuszl** [comment]
I've sent a request to API Review Board for a review, so we can go out of experimental stage for that method.

**Comment by bartonjs** [comment]
[Video](https://www.youtube.com/watch?v=eFQFAQnyLxc&t=1h4m56s)

* We feel the new property is unnecessary.  Rather, SetTimeUtc should just cease throwing the exception when time goes backwards.
  * When the new time is older, SetTimeUtc should defer to AdjustTime.
* Given no proposals exist to modify the existing AdjustTime method, we feel that it should cease being experimental (and, if there's disagreement, then the owners should come back to us).


```diff
namespace Microsoft.Extensions.Time.Testing;

public partial class FakeTimeProvider
{
-   [Experimental(diagnosticId: DiagnosticIds.Experiments.TimeProvider, UrlFormat = DiagnosticIds.UrlFormat)]
    public void AdjustTime(DateTimeOffset value)
}
```

**Comment by xackus** [comment]
>     * We feel the new property is unnecessary.  Rather, SetTimeUtc should just cease throwing the exception when time goes backwards.
>       * When the new time is older, SetTimeUtc should defer to AdjustTime.

This would make it inconsistent with `Advance(TimeSpan delta)` which also throws when trying to go back in time.
The two methods do exactly the same thing, except `SetTimeUtc` takes the target time and `Advance` the delta.

Advancing the clock vs setting the clock is meaningfully different.
`SetTimeUtc` and `Advance` simulate the passage of time.
`AdjustTime` simulates setting the system clock.

An example:
It's 5:00 PM and there is a timer due in 5 minutes.
If I SetTimeUtc(4:00 PM) the timer should now be due in 1 hour and 5 minutes.
If I AdjustTime(4:00 PM) the timer should still be due in 5 minutes.

Allowing `SetTimeUtc` and `Advance` to go back in time could make sense to simulate hardware/firmware/OS bugs.
Rust had to remove an assert that monotonic clocks don't go backwards because it kept tripping in the wild (https://doc.rust-lang.org/std/time/struct.Instant.html#monotonicity).
I don't know what .NET does on all the platforms. If `SystemTimeProvider.GetTimestamp()` is guaranteed to never go backwards even in presence of hardware/firmware/OS bugs then it might not make sense to allow this.

**Comment by amadeuszl** [comment]
I agree with @xackus regarding `SetTimeUtc()`. For now I'm just removing experimental from `AdjustTime()` as it solves the issue itself of not being able to move time back. If there's a need for a feature for going backwards and leaving timers in their place, then we can bring that.
